configure.ac

```diff
@@ -410,6 +410,29 @@
         AC_SUBST(SECLDFLAGS)
     ])
 
+    #check for plugin support
+    AC_CHECK_HEADERS([dlfcn.h])
+    AC_MSG_CHECKING([for plugin support])
+    TMPLDFLAGS="${LDFLAGS}"
+    LDFLAGS="${LDFLAGS} -rdynamic"
```

Comment:
what is -rdynamic used for?

Reply:
This is what allows the plugin to call functions inside Suricata proper.  There are other options which include:
- Break out functionality used by Suricata and plugins into a shared library. Plugins and Suricata would link with that. Snort's libdaq is like this.
- Register every symbol to export in a text file (example: https://github.com/jasonish/suricata/blob/plugins/v1/src/pluginsyms.txt)
- Provide every function in a struct of function pointers passed to the plugin on load.

Reply:
I've seen libs use a code annotation (using some macro like `PUBLIC` or `API`). Maybe that is used to generate a `pluginsyms.txt`?

Reply:
I think that is used in combination with -rdynamic to hide some symbols.  You set a default visibility, then either hide or expose.  In one example I looked at the macro became a no-op on Linux, but expanded to something that is required on Windows - something I have it looked at yet as plugins on Windows will require Suricata-core to be a shared dll from what I can tell so far.  So it has a dependency on libsuricata.

I do need to explore this area a little more.  I think we can reasonably determine what to expose for useful capture plugins, as well as file outputs.

Reply:
I think I'm edging towards limiting what we expose as opposed to defaulting to all. Even if the plugin support will be experimental for some time I think its good to already get experience with defining and maintaining a stable API, even if we'll allow for breakage during the test phase.

Reply:
Ok, I'll look at doing that.  I was playing around with -fvisibility=hidden, where we'd then use a macro to help define the function as public. This seems to affect the actual plugin, where it hides the function from Suricata - not really what we want. But doesn't have an affect on Suricata itself.

Looking at Pidgin, you do see this macro type stuff used on the API that plugins call.  Same with Snort's DAQ.  But in both these cases, the plugin is linking against a shared library, and the main program also links against that library.  So we could probably get there with libsuricata, but not now.

So we have 2 ways to control this now:
- Use -rdynamic, but control the exposed/stable API via headers and documentation.
- Or enforce what symbols are exposed - this really still requires the above, but denies a plugin from calling other functions.